targets/psa_switch/externs/psa_hash.cpp

```diff
@@ -0,0 +1,118 @@
+/* Copyright 2021 SYRMIA LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Dusan Krdzic (dusan.krdzic@syrmia.com)
+ *
+ */
+
+
+#include "psa_hash.h"
+
+namespace {
+
+std::string
+prepare_data_for_algo(const std::vector<bm::Field> &fields) {
+    bm::Data help;
+    bm::Data input;
+    static const bm::Data mask8(0xFF);
+    static const bm::Data mask4(0x0F);
+    std::string buf;
+    uint8_t n_bytes = 0;
+    uint16_t n_bits = 0;
+
+    // Concatenate fields in one single data
+    for (int i = fields.size() - 1; i >= 0; i--) {
+        help.shift_left(fields.at(i), n_bits);
+        input.add(input, help);
+        n_bits += fields.at(i).get_nbits();
+    }
+
+    // For an odd number of hex digits.
+    // 0x456 -> 0x45 0x06
+    if (n_bits % 8 != 0) {
+        help.bit_and(input, mask4);
+        buf.insert(buf.begin(), 1, help.get<uint8_t>());
+        input.shift_right(input, 4);
+        n_bits -= 4;
+    }
```

Comment:
not sure I understand this, but maybe it's because of the comment?
* comment says "For an odd number of hex digits" but then you test for an odd number of bits
* "0x456 -> 0x45 0x06" just doesn't seem like a correct transformation. I see no reason to modify the last byte in the buffer ever.

Reply:
I've found this byte extraction approach in some documentation and CRC-16 calculator on the Internet. It extracts bytes from left to right. But I agree that it doesn't seem like a correct transformation. I will remove it.